src/Microsoft.AspNetCore.WebHooks.Receivers/Extensions/WebHookRequestReaderExtensions.cs

```diff
@@ -0,0 +1,298 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.WebHooks.Utilities;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.AspNetCore.WebHooks
+{
+    /// <summary>
+    /// Extension methods for <see cref="IWebHookRequestReader"/> instances.
+    /// </summary>
+    public static class WebHookRequestReaderExtensions
+    {
+        /// <summary>
+        /// Read the HTTP request entity body as a <typeparamref name="TModel"/> instance.
+        /// </summary>
+        /// <typeparam name="TModel">The type of data to return.</typeparam>
+        /// <param name="requestReader">The <see cref="IWebHookRequestReader"/> this method extends.</param>
+        /// <param name="controllerContext">
+        /// The <see cref="ControllerContext"/> for the current request and controller.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Task"/> that on completion provides a <typeparamref name="TModel"/> instance containing the
+        /// HTTP request entity body.
+        /// </returns>
+        public static Task<TModel> ReadBodyAsync<TModel>(
+            this IWebHookRequestReader requestReader,
+            ControllerContext controllerContext)
+        {
+            if (requestReader == null)
+            {
+                throw new ArgumentNullException(nameof(requestReader));
+            }
+            if (controllerContext == null)
+            {
+                throw new ArgumentNullException(nameof(controllerContext));
+            }
+
+            return requestReader.ReadBodyAsync<TModel>(controllerContext, controllerContext.ValueProviderFactories);
+        }
+
+        /// <summary>
+        /// Read the HTTP request entity body (formatted as HTML form URL-encoded data) as an
+        /// <see cref="IFormCollection"/> instance.
+        /// </summary>
+        /// <param name="requestReader">The <see cref="IWebHookRequestReader"/> this method extends.</param>
+        /// <param name="actionContext">The <see cref="ActionContext"/> for the current request and action.</param>
+        /// <returns>
+        /// A <see cref="Task"/> that on completion provides an <see cref="IFormCollection"/> instance containing data
+        /// from the HTTP request entity body.
+        /// </returns>
+        public static async Task<IFormCollection> ReadAsFormDataAsync(
```

Comment:
What do we we need this for? As far as I know the the form data code already does all of these things

Reply:
The form data code neither ensures a re-settable stream nor resets the `Body` stream's position.

Reply:
More specifically, `ReadFormAsync(...)` doesn't always enable rewinding the stream (it's controlled by `FormOptions.BufferBody`), only drains the stream in one case, and rewinds only when the stream allows that. WebHooks needs all that.